### PR TITLE
[MXNET-560][WIP] Add temperature parameter in Softmax and SoftmaxOutput operator

### PR DESCRIPTION
**Pull request by apeforest**
## Description ##
This PR is to address the request to have a native temperature parameter in the softmax functions. See [Issue](https://github.com/apache/incubator-mxnet/issues/11016) for more detailed discussion.

I have added the temperature parameter to both softmax and SoftmaxOutput operators. By default the temperature parameter value is 1.0 and both functions should behave the same as not specifying the temperature parameter.

Verified the change using the following code in Python:

```
import mxnet as mx

data = mx.sym.Variable('data')
net = mx.sym.softmax(data=data, temperature=10)

x = mx.nd.array([ 1,  2,  3])

ex = net.bind(mx.cpu(), args={'data': x, 'softmax2_label': 'softmax2'})
ex.forward()
```
should expect return
```
[
[ 0.30060961  0.33222499  0.3671654 ]
 <NDArray 3 @cpu(0)>]
```

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- In the implementation of softmax function, I added a check for the default case where temperature equals 1.0 to remove the unnecessary divide-by-1 operation in the calculation. I have run an offline experiment of softmax function alone compiled with g++ -O3. The single core runtime can be reduced by around 25% with this optimization .

### CODE REVIEW

**Comment by apeforest** [comment]
[WIP] Only added temperature parameter to softmax but not SoftmaxOutput yet since the latter depends on operator change in another repo (MShadow). Please review the change for softmax function. Thanks

**Comment by eric-haibin-lin** [comment]
@slitsey

**Comment by apeforest** [comment]
@eric-haibin-lin Will do once I send out the PR for merge. Since this is my first contribution, I would like to get some initial review on my code change to make sure I am following the proper styles and convention in this community.

**Comment by eric-haibin-lin** [comment]
Yeah because my initial review always involves inspecting tests :)

File: src/operator/nn/softmax-inl.h [hunk 1]
```diff
@@ -145,23 +155,36 @@ __global__ void softmax_compute_kernel(DType *in, DType *out, index_t M, int axi
...
+
   __syncthreads();
   cuda::Reduce1D<red::sum, x_bits>(smem);
   __syncthreads();
   DType ssum = smem[0];
   __syncthreads();
 
-  for (index_t i = x; i < M; i += x_size) {
-    out[base + i*sa] = OP::Map(in[base + i*sa] - smax, ssum);
+  if (temperature == 1.0) {
```
Comment (rahul003) [inline src/operator/nn/softmax-inl.h]: I'd suggest removing the if condition and merging these two cases, since it's perfectly okay to use the second equation for temperature=1. The if condition within this kernel can cause branching on CPU and affect performance

File: src/operator/nn/softmax-inl.h [hunk 2]
```diff
@@ -145,23 +155,36 @@ __global__ void softmax_compute_kernel(DType *in, DType *out, index_t M, int axi
   __syncthreads();
 
   red::sum::SetInitValue(smem[x]);
-  for (index_t i = x; i < M; i += x_size) {
-    red::sum::Reduce(smem[x], static_cast<DType>(expf(in[base + i*sa] - smax)));
+  if (temperature == 1.0) {
```
Comment (rahul003) [inline src/operator/nn/softmax-inl.h]: Same here

Comment (rahul003) [reply src/operator/nn/softmax-inl.h]: Please do that everywhere possible

File: src/operator/nn/softmax-inl.h [hunk 3]
```diff
@@ -127,7 +137,7 @@ inline void SoftmaxGrad(Stream<cpu> *s, DType *out, DType *ograd,
 #ifdef __CUDACC__
 template<int x_bits, typename OP, typename DType, int ndim>
 __global__ void softmax_compute_kernel(DType *in, DType *out, index_t M, int axis,
-                                       Shape<ndim> sshape, Shape<ndim> stride) {
+                                       Shape<ndim> sshape, Shape<ndim> stride, float temperature) {
```
Comment (rahul003) [inline src/operator/nn/softmax-inl.h]: Please try to use const for variables which don't change

File: src/operator/nn/softmax-inl.h (hunk 1 above)
Comment (apeforest) [reply src/operator/nn/softmax-inl.h]: I added this if condition for performance concern. I'd assume (correct me if I'm wrong) 90% of cases the temperature to softmax is set to 1. Adding a dividep-by-1.0 operation to expf((in[base + i*sa] - smax)/t) will slow down this computation and I am not aware any compiler can optimize away this.

File: src/operator/nn/softmax-inl.h (hunk 3 above)
Comment (apeforest) [reply src/operator/nn/softmax-inl.h]: This is pass-by-value so it does not make any difference.

File: src/operator/nn/softmax-inl.h (hunk 1 above)
Comment (rahul003) [reply src/operator/nn/softmax-inl.h]: The division (even if not optimized away) should be better than causing a branch. Especially on the GPU branch divergence can have a significant impact on performance. It's better to avoid it.

File: src/operator/nn/softmax-inl.h (hunk 3 above)
Comment (rahul003) [reply src/operator/nn/softmax-inl.h]: Just a matter of style https://google.github.io/styleguide/cppguide.html#Use_of_const

Comment (apeforest) [reply src/operator/nn/softmax-inl.h]: Thanks for your suggestion. But I do not seem to find the convention of using const for pass-by-value parameter in the Google style guide: 
```
If a function guarantees that it will not modify an argument passed by reference or by pointer, the corresponding function parameter should be a reference-to-const (const T&) or pointer-to-const (const T*), respectively.
```
In fact, adding unnecessary const declaration will put restriction on the caller of the library function.

File: src/operator/nn/softmax-inl.h (hunk 1 above)
Comment (apeforest) [reply src/operator/nn/softmax-inl.h]: I agree with you about the overhead of the branching. However, there is a trade off in performance over complexity here. It really depends on how critical this piece of computation is to the overall performance of the network. In this case, I would assume softmax is a function called very often in the last layer of a neural network. But I am not an expert to oversee the performance impact.

Comment (rahul003) [reply src/operator/nn/softmax-inl.h]: If you've tested or benchmarked it, then okay. Just noticed the comment in your post.

Comment (haojin2) [reply src/operator/nn/softmax-inl.h]: @apeforest Could you provide data on which version is faster? We have `test_speed` function: https://github.com/apache/incubator-mxnet/blob/a7952f0b3218363a9520aa606f43db94a34c55b8/python/mxnet/test_utils.py#L1133 ready for you to test the speed.

File: src/operator/nn/softmax-inl.h (hunk 3 above)
Comment (haojin2) [reply src/operator/nn/softmax-inl.h]: adding `const` qualifier could be a safe net by explicitly telling the compiler about your assumption that you're not changing the input value at all in the function.
Say if you have a function like this:
```c++
void foo(int a) {
  return a;
}
```
Here your assumption that the original input should be returned without any changes, and that is essential for correct behavior of this function.
Now if someone happens to change the function to:
```c++
void foo(int a) {
  a++;      // <- new code that happens to change the value of a, and will affect correctness
  return a;
}
```
compiler will not complain about it as you do not have a `const` qualifier here, so an extra `const` qualifier is not necessary, but it could be helpful.

Comment (apeforest) [reply src/operator/nn/softmax-inl.h]: I have added const qualifier following your suggestion.

**Comment by apeforest** [comment]
@eric-haibin-lin unit test added this time :)

**Comment by haojin2** [comment]
@apeforest please do a rebase properly, you're including a lot of committed changes.

**Comment by eric-haibin-lin** [comment]
Moved to #11466
